Support using pytest-trio and hypothesis together

I'm a big fan of `hypothesis` for property-based testing. It provides a decorator, `@given`, that you use to pass in placeholders for randomly generated values. However, when I use it `@pytest.mark.trio` I get the following error:

```python
Failed: test function `<Function 'test_my_test_func'>` is marked trio but is not async
```

This happens regardless of the order that you apply the decorators. Is it possible to do something like this right now?

I'm not sure using `trio_test` works properly in this context. I have an `async` test function that takes two normal fixtures, `config_port1` and `config_port2`, and an `@given` from hypothesis.

```python
@given(msg_sent=some_strategy)
@trio_test
async def test_can_await_newline_terminated_msg(config_port1, config_port2, msg_sent):
    # Test body ...
```
This is the error message I get for that test:

```
==================================================================== test session starts =====================================================================
platform darwin -- Python 3.6.5, pytest-3.5.0, py-1.5.3, pluggy-0.6.0
rootdir: /Users/zmitchell/Projects/mhz-equip, inifile:
plugins: trio-0.4.1, hypothesis-3.55.6
collected 6 items

tests/lockin_test.py F.....                                                                                                                            [100%]

========================================================================== FAILURES ==========================================================================
___________________________________________________________ test_can_await_newline_terminated_msg ____________________________________________________________

kwargs = {}

    @wraps(fn)
>   def wrapper(**kwargs):

../../.local/share/virtualenvs/mhz-equip-IHw2MkON/lib/python3.6/site-packages/trio/testing/_trio_test.py:19:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../.local/share/virtualenvs/mhz-equip-IHw2MkON/lib/python3.6/site-packages/hypothesis/core.py:583: in execute
    result = self.test_runner(data, run)
../../.local/share/virtualenvs/mhz-equip-IHw2MkON/lib/python3.6/site-packages/hypothesis/executors.py:58: in default_new_style_executor
    return function(data)
../../.local/share/virtualenvs/mhz-equip-IHw2MkON/lib/python3.6/site-packages/hypothesis/core.py:575: in run
    return test(*args, **kwargs)
../../.local/share/virtualenvs/mhz-equip-IHw2MkON/lib/python3.6/site-packages/trio/testing/_trio_test.py:19: in test_can_await_newline_terminated_msg
    def wrapper(**kwargs):
../../.local/share/virtualenvs/mhz-equip-IHw2MkON/lib/python3.6/site-packages/hypothesis/core.py:522: in test
    result = self.test(*args, **kwargs)
../../.local/share/virtualenvs/mhz-equip-IHw2MkON/lib/python3.6/site-packages/trio/_core/_result.py:119: in unwrap
    raise self.error
../../.local/share/virtualenvs/mhz-equip-IHw2MkON/lib/python3.6/site-packages/contextvars/__init__.py:66: in run
    return callable(*args, **kwargs)
../../.local/share/virtualenvs/mhz-equip-IHw2MkON/lib/python3.6/site-packages/trio/_core/_run.py:934: in init
    async_fn, args, system_nursery, None
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = Runner(clock=SystemClock(offset=87411.59816923397), instruments=[], io_manager=KqueueIOManager(_kqueue=<select.kqueue ...read.RLock object owner=0 count=0 at 0x1022d0450>), trio_token=None, ki_pending=False, waiting_for_idle=SortedDict({}))
async_fn = functools.partial(<function test_can_await_newline_terminated_msg at 0x10215e2f0>, msg_sent=''), args = ()
nursery = <trio._core._run.Nursery object at 0x102405fd0>, name = None

    def spawn_impl(
        self, async_fn, args, nursery, name, *, ki_protection_enabled=False
    ):

        ######
        # Make sure the nursery is in working order
        ######

        # This sorta feels like it should be a method on nursery, except it
        # has to handle nursery=None for init. And it touches the internals of
        # all kinds of objects.
        if nursery is not None and nursery._closed:
            raise RuntimeError("Nursery is closed to new arrivals")
        if nursery is None:
            assert self.init_task is None

        ######
        # Call the function and get the coroutine object, while giving helpful
        # errors for common mistakes.
        ######

        def _return_value_looks_like_wrong_library(value):
            # Returned by legacy @asyncio.coroutine functions, which includes
            # a surprising proportion of asyncio builtins.
            if inspect.isgenerator(value):
                return True
            # The protocol for detecting an asyncio Future-like object
            if getattr(value, "_asyncio_future_blocking", None) is not None:
                return True
            # asyncio.Future doesn't have _asyncio_future_blocking until
            # 3.5.3. We don't want to import asyncio, but this janky check
            # should work well enough for our purposes. And it also catches
            # tornado Futures and twisted Deferreds. By the time we're calling
            # this function, we already know something has gone wrong, so a
            # heuristic is pretty safe.
            if value.__class__.__name__ in ("Future", "Deferred"):
                return True
            return False

        try:
>           coro = async_fn(*args)
E           TypeError: test_can_await_newline_terminated_msg() missing 2 required positional arguments: 'config_port1' and 'config_port2'

../../.local/share/virtualenvs/mhz-equip-IHw2MkON/lib/python3.6/site-packages/trio/_core/_run.py:756: TypeError
```

As you can see, the fixtures don't get passed to the test function. The test looks like this:

--- a/.gitignore
+++ b/.gitignore
@@ -40,6 +40,8 @@ htmlcov/
 .cache
 nosetests.xml
 coverage.xml
+.hypothesis/
+.pytest_cache/
 
 # Translations
 *.mo

--- a/newsfragments/42.feature.rst
+++ b/newsfragments/42.feature.rst
@@ -1,0 +1,2 @@
+pytest-trio now integrates with `Hypothesis <https://hypothesis.readthedocs.io>`_
+to support ``@given`` on async tests using Trio.

--- a/pytest_trio/_tests/test_hypothesis_interaction.py
+++ b/pytest_trio/_tests/test_hypothesis_interaction.py
@@ -1,0 +1,22 @@
+import pytest
+from hypothesis import given, strategies as st
+
+
+@given(st.integers())
+@pytest.mark.trio
+async def test_mark_inner(n):
+    assert isinstance(n, int)
+
+
+@pytest.mark.trio
+@given(st.integers())
+async def test_mark_outer(n):
+    assert isinstance(n, int)
+
+
+@pytest.mark.parametrize('y', [1, 2])
+@given(x=st.none())
+@pytest.mark.trio
+async def test_mark_and_parametrize(x, y):
+    assert x is None
+    assert y in (1, 2)

--- a/pytest_trio/plugin.py
+++ b/pytest_trio/plugin.py
@@ -25,8 +25,18 @@ def pytest_configure(config):
     )
 
 
-def _trio_test_runner_factory(item):
-    testfunc = item.function
+def _trio_test_runner_factory(item, testfunc=None):
+    testfunc = testfunc or item.function
+
+    if getattr(testfunc, '_trio_test_runner_wrapped', False):
+        # We have already wrapped this, perhaps because we combined Hypothesis
+        # with pytest.mark.parametrize
+        return testfunc
+
+    if not iscoroutinefunction(testfunc):
+        pytest.fail(
+            'test function `%r` is marked trio but is not async' % item
+        )
 
     @trio_test
     async def _bootstrap_fixture_and_run_test(**kwargs):
@@ -58,6 +68,7 @@ def _trio_test_runner_factory(item):
         if user_exc:
             raise user_exc
 
+    _bootstrap_fixture_and_run_test._trio_test_runner_wrapped = True
     return _bootstrap_fixture_and_run_test
 
 
@@ -215,11 +226,18 @@ def _install_async_fixture_if_needed(fixturedef, request):
 @pytest.hookimpl(hookwrapper=True)
 def pytest_runtest_call(item):
     if 'trio' in item.keywords:
-        if not iscoroutinefunction(item.obj):
-            pytest.fail(
-                'test function `%r` is marked trio but is not async' % item
+        if hasattr(item.obj, 'hypothesis'):
+            # If it's a Hypothesis test, we go in a layer.
+            item.obj.hypothesis.inner_test = _trio_test_runner_factory(
+                item, item.obj.hypothesis.inner_test
             )
-        item.obj = _trio_test_runner_factory(item)
+        elif getattr(item.obj, 'is_hypothesis_test', False):
+            pytest.fail(
+                'test function `%r` is using Hypothesis, but pytest-trio '
+                'only works with Hypothesis 3.64.0 or later.' % item
+            )
+        else:
+            item.obj = _trio_test_runner_factory(item)
 
     yield
 

--- a/test-requirements.txt
+++ b/test-requirements.txt
@@ -1,2 +1,3 @@
 pytest
 pytest-cov
+hypothesis>=3.64

